docs/design/002-containerized-gadgets.md

```diff
@@ -0,0 +1,602 @@
+
+The gadgets we provide are heavily coupled to Inspektor Gadget as they are all developed on the same
+repository and maintained by our team. Adding new gadgets implies that they have to be reviewed and
+approved by us. This approach presents some disadvantages:
+- Users need to install new versions of Inspektor Gadget to get new gadgets (or updates to them).
+- Too specific gadgets can't be added to our code base.
+- Users can't have private gadgets.
+
+The [containerized gadgets idea](https://github.com/inspektor-gadget/inspektor-gadget/issues/1669)
+aims to make Inspektor Gadget a framework to run eBPF programs (gadgets), like Docker does for
+containers. This approach decouples the gadgets from Inspektor Gadget solving the issues mentioned
+above.
+
+Many ideas of this design document are based on [bumblebee](https://github.com/solo-io/bumblebee)
+and [ebpf_exporter](https://github.com/cloudflare/ebpf_exporter) projects.
+## User Experience
+
+The UX we provide should be very close to the existing container runtimes like Docker.
+
+### `run`
+
+The run command starts a gadget:
+
+```bash
+$ ig run mygadgetimage:latest
+# print output to terminal
+```
+
+`mygadgetimage:latest` is gadget OCI image.
+This command should provide a flag to run on detached mode:
+
+```bash
+$ ig run mygadgetimage:latest --detached
+Created gadget foo
+```
+
+This command automatically pulls the image if it's available in the local image cache.
+
+### `list`
+
+List print the running gadgets
+
+```bash
+$ ig list
+...
+```
+
+### `attach`
+
+Attach to a running gadget
+
+```bash
+$ ig attach mygadget
+# gets terminal attached to the gadget
+```
+
+### `pull`
+
+Pulls a gadget image
+
+```bash
+$ ig pull mygadgetimage:latest
+```
+
+### `build`
+
+Builds and packages a gadget as an OCI image
+
+```bash
+$ ig build --prog foo.bpf.c --definition foo.yaml mygadgetimage:latest
+```
+
+Details of this command are still to be defined. Like other flags that are supported.
+
+### `tag`
+
+Tag an image
+
+```bash
+$ ig tag mygadgetimage:latest ghcr.io/foo/mygadgetimage:v1
+```
+
+### `image`
+
+Different commands to handle images
+
+- `image list`
+- `image delete`
+etc...
+
+TODO: Should commands like `ig pull`, `ig build`, `ig push` be aliases of `ig image ...` ones?
+
+## Gadget Packaging
+
+Gadgets should be packed as [OCI](https://github.com/opencontainers/image-spec) images. Initially
+the following layers are proposed:
+- `application/config.image.inspektor-gadget.io+yaml`: [Configuration file](#configuration-file)
+- `application/ebpf.image.inspektor-gadget.io+binary`: Compiled eBPF object.
+
+### Configuration File
+
+The configuration file contains the following information for a gadget:
+- `name`: Human readable name of the gadget, like `trace_exec`, `profile_ebpf`, etc.
+- `description`: Short description about the gadget.
+- `documentation`: Long text describing how to use the gadget. This should describe the different
```

Comment:
Some of those fields are redundant with the OCI annotations that are already defined for all OCI images in:
https://github.com/opencontainers/image-spec/blob/main/annotations.md#pre-defined-annotation-keys

So instead of duplicating them (and then we don't know which ones are the valid ones), should we give a recommendation that users should set the OCI annotations like documentation, url, etc.?

If we need to duplicate them for practical reasons, I would like to have a generic `annotations map[string]string` reusing the keys defined by OCI.

Reply:
Yes, I agree we should use the OCI ones. Also because the configuration file should be optional later on.